src/python/pants/backend/python/util_rules/pex.py

```diff
@@ -421,19 +427,44 @@ async def build_pex(
     requirements_file_digest = EMPTY_DIGEST
     if request.requirements.file_path:
         argv.extend(["--requirement", request.requirements.file_path])
+
+        globs = PathGlobs(
+            [request.requirements.file_path],
+            glob_match_error_behavior=GlobMatchErrorBehavior.error,
+            description_of_origin=request.requirements.file_path_description_of_origin,
+        )
+
+        # use contents to invalidate here
+        requirements_file_digest_contents = await Get(
+            DigestContents,
+            PathGlobs,
+            globs,
+        )
+        metadata = read_lockfile_metadata(requirements_file_digest_contents[0].content)
+        if metadata.invalidation_digest != request.requirements.lockfile_hex_digest:
+            if python_setup.fail_on_invalid_lockfile:
+                raise ValueError("Invalid lockfile provided. [TODO: Improve message]")
+            else:
+                logger.warning("%s", "Invalid lockfile provided. [TODO: Improve message]")
+
         requirements_file_digest = await Get(
             Digest,
-            PathGlobs(
-                [request.requirements.file_path],
-                glob_match_error_behavior=GlobMatchErrorBehavior.error,
-                description_of_origin=request.requirements.file_path_description_of_origin,
-            ),
+            PathGlobs,
+            globs,
         )
+
     elif request.requirements.file_content:
-        argv.extend(["--requirement", request.requirements.file_content.path])
-        requirements_file_digest = await Get(
-            Digest, CreateDigest([request.requirements.file_content])
-        )
+        content = request.requirements.file_content
+        argv.extend(["--requirement", content.path])
+
+        metadata = read_lockfile_metadata(content.content)
+        if metadata.invalidation_digest != request.requirements.lockfile_hex_digest:
+            if python_setup.fail_on_invalid_lockfile:
+                raise ValueError("Invalid lockfile provided. [TODO: Improve message]")
+            else:
+                logger.warning("%s", "Invalid lockfile provided. [TODO: Improve message]")
```

Comment:
We will likely want this error message to differ from the one above. This is currently always the `<default>` lockfile, whereas the above are always lockfiles checked in by the user. Our guidance is different depending on those two edge cases.

To formalize that implicit association, we may in a followup want to change `PexRequirements` to be less generic than it is now, e.g. change the field to be "default_requirements_file: FileContent` rather than `file_content: FileContent`.

Reply:
Addressing in a follow-up

Reply:
Okay, fair enough. I do think we have some tests using this style out there and iirc Benjy prefers your style - not blocking